src/main/java/net/kaczmarzyk/spring/data/jpa/web/WebRequestQueryContext.java

```diff
@@ -51,13 +53,44 @@ public WebRequestQueryContext(NativeWebRequest request) {
 			this.evaluatedJoinFetch = new HashMap<>();
 			request.setAttribute(JOIN_FETCH_ATTRIBUTE_KEY, evaluatedJoinFetch, NativeWebRequest.SCOPE_REQUEST);
 		}
+                
+                this.rootCache = (Map<Pair<String, Root>, javax.persistence.criteria.Join<?, ?>>) request.getAttribute(ROOT_CACHE_ATTRIBUTE_KEY, NativeWebRequest.SCOPE_REQUEST);
+
+		if (this.rootCache == null) {
+			this.rootCache = new HashMap<>();
+			request.setAttribute(ROOT_CACHE_ATTRIBUTE_KEY, rootCache, NativeWebRequest.SCOPE_REQUEST);
+		}
 	}
 
+        @Override
+        public boolean existsJoin(String key, Root<?> root) {
+            Function<Root<?>, Join<?, ?>> value = contextMap.get(key);
+
+            if (value == null) {
+                    Optional<? extends Join<?, ?>> opJoin = root.getJoins().stream().filter(p -> p.getAlias() != null && p.getAlias().equals(key)).findFirst();
+                    if(opJoin.isPresent()){
+                        Join<?, ?> join = opJoin.get();
+                        putLazyVal(key, (r) -> join);
+                        return true;
+                    }
+                    return false;
+            }
+            return true;
+        }
+        
+        
+
 	@Override
 	public Join<?, ?> getEvaluated(String key, Root<?> root) {
 		Function<Root<?>, Join<?, ?>> value = contextMap.get(key);
 
 		if (value == null) {
+                        Optional<? extends Join<?, ?>> opJoin = root.getJoins().stream().filter(p -> p.getAlias() != null && p.getAlias().equals(key)).findFirst();
+                        if(opJoin.isPresent()){
+                            Join<?, ?> join = opJoin.get();
+                            putLazyVal(key, (r) -> join);
+                            return join;
+                        }
```

Comment:
Is it necessary? (As far as I understand the specification-argument-resolver mechanisms, each join should be placed in context map). In which situation the `join` with given alias will be not present in `contextMap` and it will be present in `root.rootJoins()` ?

Reply:
Really, I don't remember anymore, I'll remove the code

Reply:
I got confused, the code that was in existsJoin I don't remember. This one is for when the person doesn't add the first Join. Put the join table1.table2 directly